### DISCUSSION TITLE
Is it possible to use closed source pylance?

### DISCUSSION THREAD

**Discussion by itsfarseen** [emacs-lsp/lsp-mode#2996] [General]
If I have a VSCode installation, is it possible to copy over the pylance executable and somehow make it work with lsp-mode?

**Comment by itsfarseen** [comment]
Motivation:
A lot of completions which were being shown in vim, is not shown in emacs (pyright). Notably, field names of typed dicts.

I am a long time Vim user. I use pylance running on top of Coc. Installed Doom Emacs last week and loving it so far. 
It would be really nice if using pylance with emacs is possible.

**Reply by yyoncho** [reply]
pylance license does not allow running it outside of vscode/ms products.

**Reply by itsfarseen** [reply]
I was going to post but coc can run it, but I double checked and it's using pyright only :smile:  
But coc-pyright autocompletes field names of TypedDicts, while emacs doesn't. What could be causing this?

**Reply by itsfarseen** [reply]
coc-pyright also have auto imports. So I was under the impression that they are actually using pylance.

**Reply by yyoncho** [reply]
> I was going to post but coc can run it, but I double checked and it's using pyright only 
> But coc-pyright autocompletes field names of TypedDicts, while emacs doesn't. What could be causing this?

a lot of things  - different default settings, different root selection, missing logic on lsp-mode side. In order to determine the difference it will be great if you can provide sample project + steps to reproduce. I will crosscheck with vscode.

**Reply by dsyzling** [reply]
@itsfarseen what is your expected behaviour? If I define a simple test from the TypedDict pep:
```python
from typing import TypedDict


def test_typed_dict() -> None:
    class Movie(TypedDict):
        name: str
        year: int

    movie: Movie = {'name': 'Blade Runner',
                    'year': 1982}

    assert(movie['year'] == 1982)
    assert(movie['name'] == "Blade Runner")
```
When I type movie[  - pyright prompts me with the defined attributes for the TypedDict class - i.e. name, year. Equally pyright validates the types of the Movie fields when constructing the 'movie' dictionary instance - i.e. if you pass a string for year.

**Reply by itsfarseen** [reply]
Ooh there must be something wrong with my config then.

No type error: (Left doom emacs, right neovim + coc) 
![image](https://user-images.githubusercontent.com/40014472/126185526-d44100d2-1296-4b6a-a4ce-aaf0e035b617.png)

No field name completion: 
![image](https://user-images.githubusercontent.com/40014472/126185919-2e401d06-904c-4b58-8c9d-35500cd5b39c.png)

If I press control space when inside the quotes, emacs hangs for a few mins.

**Reply by itsfarseen** [reply]
Yupp. I was not using `+pyright`, I only had `(python +lsp)`

**Reply by dsyzling** [reply]
Ah - so everything works now?

**Reply by itsfarseen** [reply]
It works after I added `+pyright` and did `SPC d r r`. Thanks @dsyzling

**Comment by Mariownyou** [comment]
I am sorry but what colorsheme do you use?

**Reply by pmbittner** [reply]
That is "doom-one".
